src/opustrainer/trainer.py

```diff
@@ -24,6 +28,101 @@ def ignore_sigint():
     """
     signal.signal(signal.SIGINT, signal.SIG_IGN)
 
+# Placeholders, the Chinese edition, with some hardcoded values
+
+md = MosesDetokenizer(lang='zh')
+
+def get_placeholding_candidates(align_line: str) -> List[Tuple[int, int]]:
+    """Filters out multiple alignment targets so that we can definitely get a one-to-one replacement"""
+    def tuplify(pair: str) -> Tuple[int, int]:
+        s, t = pair.split('-')
+        return (int(s), int(t))
+    # Create the two src-trg and trg-src sets
+    src_trg: List[Tuple[int, int]] = [tuplify(i) for i in align_line.split()]
+    trg_src: List[Tuple[int, int]] = [(c, d) for d,c in src_trg]
+    # Sort them
+    src_trg.sort(key=lambda x: x[0])
+    trg_src.sort(key=lambda x: x[0])
+    # Remove anything that is found multiple times. Anything that is found multiple times means non bijective alignment
+    # Since they are sorted, we can reduce some time complexity
+    def filter_tuples(inlist: List[Tuple[int, int]]) -> List[Tuple[int, int]]:
+        '''Removes places non x->y x->z type of alignments'''
+        new_list: List[Tuple[int, int]] = []
+        blacklisted: Set = set()
+        for i in range(len(inlist)):
+            curr: int = inlist[i][0]
+            if curr not in blacklisted:
+                if i < len(inlist) - 1:
+                    peek: int = inlist[i+1][0]
+                    if curr==peek:
+                        blacklisted.add(curr)
+                    else:
+                        new_list.append(inlist[i])
+                else:
+                    new_list.append(inlist[i])
+        return new_list
+
+    src_trg_filtered: List[Tuple[int, int]] = filter_tuples(src_trg)
+    trg_src_filtered: List[Tuple[int, int]] = filter_tuples(trg_src)
+
+    # Now, we are looking for the union of the two.
+    # First, reverse src_trg
+    trg_src_filtered_rereversed: List[Tuple[int, int]] = [(c, d) for d,c in trg_src_filtered]
+
+    # Now convert both to sets and take the union
+    src_trg_set: Set[Tuple[int, int]] = set(src_trg_filtered)
+    trg_src_filtered_rereversed_set: Set[Tuple[int, int]] = set(trg_src_filtered_rereversed)
+
+    return list(src_trg_set & trg_src_filtered_rereversed_set)
+
+# Unpacks a line, removes the alignments, and applies placeholding. Hardcoded for the moment
+# Also applies detokenization on the source side, because getting word alignments for Chinese is otherwise hard
+
+def tag_line(line: str, *, probability: float=0.0, num_tags: int=6, chinese: str=None) -> str:
```

Comment:
How about `str | None=None` ?

Reply:
Yeah i gave up on the enum idea a while ago.... We should change the  requirement of the build from 3.7 to 3.9 then since we're not compatible with 3.8.